### ISSUE TITLE
Request: wrap current form in list and enter insert mode at head or tail

### ISSUE THREAD

**Issue by NoahTheDuke** [julienvincent/nvim-paredit#35]
Hey friends,

vim-sexp has a bunch of ["wrap current list in new X and enter insert mode at head/tail"](https://github.com/guns/vim-sexp/blob/14464d4580af43424ed8f2614d94e62bfa40bb4d/plugin/sexp.vim#L67-L72) actions. I frequently find myself wanting to wrap a form that produces a seq (in Clojure) and filter it, so I type `\ifilter some-pred`, which handles adding the wrapping list and putting me directly into insert mode to add the necessary function call. This can be accomplished already by using a surround plugin to select the current form and then add the surrounding brackets and then enter insert mode myself, but comparatively (with nvim-surround) I have to press `va)S)` and then either press `i` for the head or `%i` to reach the tail and enter insert mode.

(vim-sexp also includes wrapping in `[`/`]` and `{`/`}`, but that can be done once the initial work to build lists is done, which I suspect is the majority of cases.)

I don't know if this would be hard to implement or not, but I would greatly appreciate its inclusion.

Thank you!

**Comment by julienvincent** [comment]
Hey, I think what you are wanting to achieve can already be fully accomplished with our wrap + cursor API's.

See https://github.com/julienvincent/nvim-paredit?tab=readme-ov-file#vim-sexp-wrap-form-headtail-replication

We don't offer any default keymaps for this and stherefore it's a little more manual to configure, but I think it should do what you want.

Let me know if this isn't what you were looking for.

**Comment by NoahTheDuke** [comment]
That looks like what I'm hoping to add, yes. Thank you.

As I've got you here, I'm hoping to mimic the `<localleader>h` and `<localleader>l` functionality as well, that jumps to insert mode within the current list/vector/etc at the start or end. I wrote the below based on `wrap_enclosing_form_under_cursor` but I'm not sure if it's the right way to do it.

```lua
local function insert_in_list(placement)
  return function()
    local lang = langs.get_language_api()
    local current_element = wrap.find_element_under_cursor(lang)

    if not current_element then
      return
    end

    local use_direct_parent = common.is_whitespace_under_cursor(lang)
      or lang.node_is_comment(ts.get_node_at_cursor())

    local form = wrap.find_form(current_element, lang)
    if not use_direct_parent and form:type() ~= "source" then
      form = wrap.find_parend_form(current_element, lang)
    end

    return paredit.cursor.place_cursor(
      form,
      { placement = placement, mode = "insert" }
    )
  end
end

paredit.setup({
  keys = { 
    -- existing stuff

    ["<localleader>h"] = {
      insert_in_list("inner_start"),
      "Enter insert mode at head of current form"
    },
    ["<localleader>l"] = {
      insert_in_list("inner_end"),
      "Enter insert mode at tail of current form"
    },
```

**Comment by julienvincent** [comment]
At the moment we don't necessarily have a right way to do it haha. I think what you have produced is good but keep in mind that you are consuming some internals we haven't explicitly made public (yet) and so I can't promise we won't change them down the line.

But I do want to make these internals public at some point to allow for this kind of scriptability, at least in a more officially supported way!

@armed might be able to weigh in here a bit. Do we want to add some more API's to better support this use case?

**Comment by armed** [comment]
Yep, we definitely need to expand API. Just need to be careful with a shape of it to avoid future breaking changes.

**Comment by armed** [comment]
hey, @NoahTheDuke 
Sorry for replying so late. I believe wrapping list with stuff and inserting cursor actually can be already done with an existing API:

```lua
      ["<localleader>wt"] = {
        function()
          paredit.wrap.wrap_enclosing_form_under_cursor("(doto ", " tap>)")
        end,
        "Wrap with doto tap>",
      },

      ["<localleader>i"] = {
        function()
          paredit.cursor.place_cursor(
            paredit.wrap.wrap_enclosing_form_under_cursor("( ", ")"),
            { placement = "inner_start", mode = "insert" }
          )
        end,
        "Wrap form insert head",
      },

      ["<localleader>I"] = {
        function()
          paredit.cursor.place_cursor(
            paredit.wrap.wrap_enclosing_form_under_cursor("(", ")"),
            { placement = "inner_end", mode = "insert" }
          )
        end,
        "Wrap form insert tail",
      },

```

example:
```clojure
(+ 1| 1)
```
press `<localleader>wt`:
```
(doto (+ 1| 1) tap>)
```

Also for wrapping with curly and square brackets:
```lua
      ["<localleader>e["] = {
        function()
          paredit.cursor.place_cursor(
            paredit.wrap.wrap_element_under_cursor("[", "]"),
            { placement = "inner_start", mode = "insert" }
          )
        end,
        "Wrap form square insert head",
      },
      ["<localleader>e]"] = {
        function()
          paredit.cursor.place_cursor(
            paredit.wrap.wrap_element_under_cursor("[", "]"),
            { placement = "inner_end", mode = "insert" }
          )
        end,
        "Wrap form square insert tail",
      },

      ["<localleader>e{"] = {
        function()
          paredit.cursor.place_cursor(
            paredit.wrap.wrap_element_under_cursor("{", "}"),
            { placement = "inner_start", mode = "insert" }
          )
        end,
        "Wrap form curly insert head",
      },
      ["<localleader>e}"] = {
        function()
          paredit.cursor.place_cursor(
            paredit.wrap.wrap_element_under_cursor("{", "}"),
            { placement = "inner_end", mode = "insert" }
          )
        end,
        "Wrap form curly insert tail",
      },
```

is this what you looking for?

**Comment by armed** [comment]
what we are missing though is `unwrap` API, which is I'm going to implement soon

**Comment by julienvincent** [comment]
> what we are missing though is `unwrap` API, which is I'm going to implement soon

Which we now have thanks @armed :) I think this can probably be closed out then, or is there still something left to action?

**Comment by armed** [comment]
Yep, can be closed

**Comment by NoahTheDuke** [comment]
Apologies for my own late reply lol. I tried out your suggested code and then forgot to comment! Two thoughts:

1) `paredit.wrap.wrap_element_under_cursor` is great except that it currently doesn't work when over whitespace: `(1 2| 3)` -> `(| (1 2 3))` but `(1 2 |3)` -> `(1 2 |3)` Is there any way to fix this? I suspect it has to do with `is_whitespace_under_cursor` or something, but I've not dug in deep enough.

2) I've modified your recommended `<leader>i` and `<leader>I` mappings to work on the current form if directly on it. I don't know if this is something you want to replicate or support, but I find it very annoying to have to move inside of a form to be able to wrap it, instead of having that happen as I anticipate.
    * With your suggested binding: `(1 2 (|3 4) 5)` -> `(| (1 2 (3 4) 5))`
    * With my changes: `(1 2 (|3 4) 5)` -> `(1 2 (| (3 4)) 5)`

```lua
local function enclosing_wrapper_maker(brackets, placement)
  return function()
    -- If the node is a "form", then treat it like an element: (|1 2 3) -> (| (1 2 3))
    if clj.node_is_form(ts.get_node_at_cursor()) then
      paredit.cursor.place_cursor(
        paredit.wrap.wrap_element_under_cursor(unpack(brackets)),
        { placement = placement, mode = "insert" }
      )
    -- Otherwise, wrap the surrounding form: (1 2| 3) -> (| (1 2 3))
    else
      paredit.cursor.place_cursor(
        paredit.wrap.wrap_enclosing_form_under_cursor(unpack(brackets)),
        { placement = placement, mode = "insert" }
      )
    end
  end
end

paredit.setup({
  ...
  keys = {
    ["<localleader>i"] = {
      enclosing_wrapper_maker({"( ", ")"}, "inner_start"),
      "Wrap form round, insert head",
    },
    ["<localleader>I"] = {
      enclosing_wrapper_maker({"(", " )"}, "inner_end"),
      "Wrap form round, insert tail",
    },
  ...
  }
})
```

**Comment by armed** [comment]
@NoahTheDuke that's a valid point. Personally I'm pretty comfortable with both solutions (not a heavy user of that keybindings though, so no muscle memory), if @julienvincent doesn't have anything against it then I'll advice you to make a PR :)

**Comment by armed** [comment]
Oh wait, those keybindings not in the defaults list - not the part of a plugin, they are described in the recipes section. So we can just update recipes section.

**Comment by julienvincent** [comment]
To be honest, I think these keybindings should probably just be added as defaults to nvim-paredit. @armed any objection to that?

**Comment by NoahTheDuke** [comment]
With the recent updates to nvim-paredit now relying on treesitter queries, what's the best way to update the code I have?

**Comment by julienvincent** [comment]
Hey - sorry for breaking your workflow!

Assuming the only internals you were using is the ones listed here: https://github.com/julienvincent/nvim-paredit/issues/35#issuecomment-1735739956 then you should be able to alter it as such:

```lua
local ts_context = require("nvim-paredit.treesitter.context")
local ts_forms = require("nvim-paredit.treesitter.forms")
local ts_utils = require("nvim-paredit.treesitter.utils")
local whitespace = require("nvim-paredit.api.whitespace")

local function insert_in_list(placement)
  return function()
    local context = ts_context.create_context()
    if not context then
      return
    end

    local current_element = ts_forms.get_node_root(context.node, context)
    if not current_element then
      return
    end

    local use_direct_parent = whitespace.is_whitespace_under_cursor() or ts_utils.node_is_comment(current_element, context)

    local form = ts_forms.find_nearest_form(current_element, {
      captures = context.captures,
      use_source = false,
    })
    if not form then
      return
    end

    if not use_direct_parent and form:type() ~= "source" then
      form = wrap.find_parent_form(current_element, context)
    end

    return paredit.cursor.place_cursor(form, { placement = placement, mode = "insert" })
  end
end
```

Let me know if that doesn't work for you.

I also realized I need to update the recipes in the docs which are still using the old API.

---

This is all the more reason to make this usage a public API I suppose.
